components/oauth-client/src/client.rs

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) 2018 Chef Software Inc. and/or applicable contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::env;
+use reqwest::{self, header};
+use config::OAuth2Cfg;
+use error::Result;
+use types::*;
+
+use builder_core::metrics::CounterMetric;
+use metrics::Counter;
+use github::GitHub;
+use gitlab::GitLab;
+use bitbucket::Bitbucket;
+
+pub struct OAuth2Client {
+    inner: reqwest::Client,
+    pub config: OAuth2Cfg,
+    pub provider: Box<OAuth2Provider>,
+}
+
+impl OAuth2Client {
+    pub fn new(config: OAuth2Cfg) -> Self {
+        let mut headers = header::Headers::new();
+        headers.set(header::UserAgent::new("oauth-client"));
+        let mut client = reqwest::Client::builder();
+        client.default_headers(headers);
+
+        for proxy_var in &["HTTP_PROXY", "HTTPS_PROXY"] {
+            if let Ok(url) = env::var(proxy_var) {
+                match reqwest::Proxy::http(&url) {
+                    Ok(p) => {
+                        client.proxy(p);
+                    }
+                    Err(e) => warn!("Invalid proxy url: {}, err: {:?}", url, e),
+                }
+            }
+        }
+
+        let provider: Box<OAuth2Provider> = match &config.provider[..] {
```

Comment:
Wow, I've never seen the `[..]` syntax before.  Looking at the [Rust syntax index](https://doc.rust-lang.org/stable/book/first-edition/syntax-index.html) makes it seem like that's for collection indexing?  But I don't think I follow how that works if the provider is a boxed trait object.

Reply:
This is a bit of an known issue in Rust where you can't just match a &String to a static literal - https://github.com/rust-lang/rust/issues/28606

So you have to do an explicit conversion.. the [..] syntax is basically converting the String to a string slice str (there's other ways to do it as well).

Reply:
Wow, far out.  Thanks for explaining.